File: tags/implementation-before-prop-refact/UrakawaToolkit/media/data/utilities/CollectManagedMediaTreeNodeVisitor.cs
using System;
using System.Collections.Generic;
using System.Text;
using urakawa.core;
using urakawa.core.visitor;
using urakawa.property.channel;

namespace urakawa.media.data.utilities
{
	/// <summary>
	/// Visitor that collects all <see cref="MediaData"/> used by the visited <see cref="TreeNode"/>s.
	/// </summary>
	public class CollectManagedMediaTreeNodeVisitor : ITreeNodeVisitor
	{
		private List<IManagedMedia> mCollectedMedia = new List<IManagedMedia>();

		/// <summary>
		/// Gets the list of collected <see cref="IManagedMedia"/>
		/// </summary>
		/// <returns>The list</returns>
		/// <remarks>
		/// The returned list is a reference to the list in the <see cref="CollectManagedMediaTreeNodeVisitor"/> instance, 
		/// any changes made to the returned list will reflect in 
		/// the <see cref="CollectManagedMediaTreeNodeVisitor"/> instance
		/// </remarks>
		public List<IManagedMedia> getListOfCollectedMedia()
		{
			return mCollectedMedia;
		}

		#region ITreeNodeVisitor Members

		/// <summary>
		/// Any <see cref="IManagedMedia"/> used by the 
		/// </summary>
		/// <param name="node">The node being visited</param>
		/// <returns><c>true</c></returns>
		public bool preVisit(TreeNode node)
		{
			foreach (Type propType in node.getListOfUsedPropertyTypes())
			{
				if (propType.IsSubclassOf(typeof(ChannelsProperty)))
				{
					ChannelsProperty chProp = (ChannelsProperty)node.getProperty(propType);
					foreach (Channel ch in chProp.getListOfUsedChannels())
					{
						if (chProp.getMedia(ch) is IManagedMedia)
						{
							IManagedMedia mm = (IManagedMedia)chProp.getMedia(ch);
							if (!mCollectedMedia.Contains(mm)) mCollectedMedia.Add(mm);
						}
					}
				}
			}
			return true;
		}

		/// <summary>
		/// Nothing is done in post visit
		/// </summary>
		/// <param name="node">The node being visited</param>
		public void postVisit(TreeNode node)
		{
			return;
		}

		#endregion
	}
}

File: java/core/src/org/daisy/urakawa/core/visitor/examples/CollectManagedMediaTreeNodeVisitor.java
package org.daisy.urakawa.core.visitor.examples;

import java.util.LinkedList;
import java.util.List;

import org.daisy.urakawa.core.ITreeNode;
import org.daisy.urakawa.core.visitor.ITreeNodeVisitor;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.media.data.IManagedMedia;
import org.daisy.urakawa.property.IProperty;
import org.daisy.urakawa.property.channel.IChannel;
import org.daisy.urakawa.property.channel.ChannelDoesNotExistException;
import org.daisy.urakawa.property.channel.IChannelsProperty;

/**
 * Visitor that collects all IMediaData used by the visited TreeNodes
 */
public class CollectManagedMediaTreeNodeVisitor implements ITreeNodeVisitor {
	private List<IManagedMedia> mCollectedMedia = new LinkedList<IManagedMedia>();

	/**
	 * @return list
	 */
	public List<IManagedMedia> getListOfCollectedMedia() {
		return mCollectedMedia;
	}

	public boolean preVisit(ITreeNode node) {
		for (IProperty prop : (List<IProperty>) node.getListOfProperties()) {
			if (prop instanceof IChannelsProperty) {
				IChannelsProperty chProp = (IChannelsProperty) prop;
				for (IChannel ch : chProp.getListOfUsedChannels()) {
					try {
						if (chProp.getMedia(ch) instanceof IManagedMedia) {
							IManagedMedia mm = (IManagedMedia) chProp
									.getMedia(ch);
							if (!mCollectedMedia.contains(mm))
								mCollectedMedia.add(mm);
						}
					} catch (MethodParameterIsNullException e) {
						// Should never happen
						throw new RuntimeException("WTF ??!", e);
					} catch (ChannelDoesNotExistException e) {
						// Should never happen
						throw new RuntimeException("WTF ??!", e);
					}
				}
			}
		}
		return true;
	}

	public void postVisit(@SuppressWarnings("unused")
	ITreeNode node) {
		return;
	}
}
